### PR TITLE
Support multiple funders on work

### LINKED ISSUE #717: Funder data should be multi-valued
- See also #729

We need to be able to describe funding information that looks like this (from https://dataspace.princeton.edu/handle/88435/dsp01rj4307478?mode=full): 

> The U.S. Department of Energy (DE-SC0018925,DE-AC02-76SF00515), the National Science Foundation (NSF) (EAR-1644614, EAR-1446969, EAR-1725349), and the German Research Foundation DFG (AP 262/2-1 and FOR2440).

Mock up: 

Screenshot 2022-12-12 at 6 11 11 PM

## Acceptance criteria
- [ ] Funding reference is multi-valued and a user can keep adding more funding references as needed
- [ ] All funding references appear in the DataCite record that is produced

### PR DESCRIPTION
**Pull request by mccalluc**
- Fix #717

This is simpler than the approach taken with the other multi-valued fields: Using Rails param conventions for repeating fields means we don't need JS to copy data into the table or to keep track of the row count, and the ruby form processing code is also simpler. The JS for adding a row is generic and could be used to add rows to any table. If this much is acceptable, the same patterns could be applied to other fields:
- #761 
- #760
- #765 

Questions / TODO:
- Could this much be merged, or do we need the items below?
- Besides adding rows, do we want to be able to reorder and delete rows?
- Should we use a table for display of funders even when not editing? Or something else.
- More testing.
- Make sure funder changes show up in prov log.

### CODE REVIEW

File: app/models/pdc_serialization/datacite.rb
Comment (carolyncole) [inline app/models/pdc_serialization/datacite.rb:221]: Shouldn't this be funder.funder name here?  I feel like this code must be untested.

File: app/services/form_to_resource_service.rb
Comment (carolyncole) [inline app/services/form_to_resource_service.rb:30]: I would prefer if the method returns the updated resource,  Just to make the code clear that the resource is being changed by the method.  I think that is the pattern process_contributors takes.  That said we should make it all constant.

File: app/services/form_to_resource_service.rb
Comment (carolyncole) [inline app/services/form_to_resource_service.rb:121]: I'm unsure what the params are going to look like here.  That does not indicate a code change as long as it is show in the tests.

File: app/services/form_to_resource_service.rb
Comment (carolyncole) [inline app/services/form_to_resource_service.rb:129]: This explains why the need for filter map...

File: app/views/works/_form_hidden_fields.html.erb
Comment (carolyncole) [inline app/views/works/_form_hidden_fields.html.erb:12]: Why remove this comment.  It still seems relevant

File: spec/system/work_edit_spec.rb
Comment (carolyncole) [inline spec/system/work_edit_spec.rb:135]: Need to test multiple

**Review by carolyncole — CHANGES_REQUESTED** [review]
There seems to be some untested code.  Also it would be good to test multiple items to show your new params pattern.  Additionally the javascript is entirely untested.  A system test could help that.

File: app/models/pdc_serialization/datacite.rb
Comment (mccalluc) [reply app/models/pdc_serialization/datacite.rb:221]: thanks! will do.

File: app/services/form_to_resource_service.rb
Comment (mccalluc) [reply app/services/form_to_resource_service.rb:30]: My preference would be that it either modify in place, or assign to a new instance. Right now it's doing both, and I think that could be confusing. `process` is pretty vague, so I'll rename the methods for clarity.

File: app/services/form_to_resource_service.rb
Comment (mccalluc) [reply app/services/form_to_resource_service.rb:129]: Right: This is pattern that is already used for creator and contributor, but with an extra `if` at the end. If there's an empty row in the table, we don't create a record for it.

File: app/views/works/_form_hidden_fields.html.erb
Comment (mccalluc) [reply app/views/works/_form_hidden_fields.html.erb:12]: The same hold for all the fields above: `new_title_count` tells the controller how many new titles to expect, `creator_count` tells the controller how many creators to expect, etc. 

In the slightly longer run, we won't need these fields as we move towards using the `[]` rails convention for repeated field names.

File: spec/system/work_edit_spec.rb
Comment (mccalluc) [reply spec/system/work_edit_spec.rb:135]: Thanks!

**Comment by mccalluc** [comment]
Back to draft. Need to add
- [ ] Test of serialization
- [ ] Test of data entry for multiple
- [ ] Rename `process_*` fields
- [ ] Test of prov logging

Carolyn didn't comment on these, so I think they can wait for a future PR?
- reorder and delete
- better display when not editting; table?

**Comment by mccalluc** [comment]
Probably hold off on reviewing this until #771 is reviewed and merged: It handles part of the testing, and I've merged that branch into this.

Looking at your comments, I was unsure about this:
> That said we should make it all constant.

Do you mean that immutable objects are good? Generally, I'd agree, but when it's just breaking the constructor down into smaller pieces, using private methods, I'm ok with it.

The more important thing to me is that a set of methods are consistent about whether they modify the object, or return a new object, but these methods were doing both.

File: app/views/works/show.html.erb
Comment (carolyncole) [inline app/views/works/show.html.erb:179]: Are we loosing the Headings here? `Funder Name` `Award Number` `Award URI:`

**Review by carolyncole — APPROVED** [review]
Just one question on the show page.  That said I do not think that should block this PR, but we should know what the fields represent.  Maybe ticket deciding how the Funders should be shown?

File: app/views/works/show.html.erb
Comment (mccalluc) [reply app/views/works/show.html.erb:179]: I had thought it might be clear enough from context: Each subfield has a distinct format. Since this will repeat, I'm not sure if this would be better presented as a table. 
- Filed #785 for follow-up.

**Pull request merged**
